File: ralib/src/main/java/com/app/ralib/patch/PatchManager.java
package com.app.ralib.patch;

import android.util.Log;

import androidx.annotation.Nullable;

import com.app.ralib.Shared;
import com.app.ralib.extractors.BasicSevenZipExtractor;
import com.app.ralib.extractors.ExtractorCollection;
import com.app.ralib.utils.FileUtils;
import com.app.ralib.utils.TemporaryFileAcquirer;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class PatchManager {
    private static final String TAG = "PatchManager";

    private static final boolean IS_DEFAULT_PATCH_STORAGE_DIR_EXTERNAL = true;
    private static final String PATCH_STORAGE_DIR = "patches";

    private final Path patchStoragePath;
    private final Path configFilePath;
    private PatchManagerConfig config;

    public PatchManager(@Nullable String customStoragePath) throws IOException {
        this(customStoragePath, false);
    }

    public PatchManager(@Nullable String customStoragePath, boolean installPatchesImmediately) throws IOException {
        var isFirstTime = false;
        patchStoragePath = getDefaultPatchStorageDirectories(customStoragePath);
        if (!Files.isDirectory(patchStoragePath) || !Files.exists(patchStoragePath)) {
            Files.deleteIfExists(patchStoragePath);
            Files.createDirectories(patchStoragePath);
            isFirstTime = true;
        }
        configFilePath = patchStoragePath.resolve(PatchManagerConfig.CONFIG_FILE_NAME);
        loadConfig();

        // 如果指定立即安装，则在当前线程安装补丁（用于向后兼容）
        // 否则，应该由调用者在后台线程调用 installBuiltInPatches
        if (installPatchesImmediately) {
            installBuiltInPatches(this);
        }
    }

    //region Patch Querying

    /**
     * Returns all patches that are applicable to the specified game and are enabled for
     * the provided game assembly path. Results are sorted by priority in descending order
     * (higher priority first).
     * 
     * Note: By default, ALL applicable patches are enabled. Only explicitly disabled patches
     * will be filtered out.
     *
     * @param gameId      the game ID to filter patches by
     * @param gameAsmPath the path to the game's assembly (used to lookup enabled patches)
     * @return a list of applicable and enabled patches, sorted by priority (highest first)
     */
    public ArrayList<Patch> getApplicableAndEnabledPatches(String gameId, Path gameAsmPath) {
        var installedPatches = getInstalledPatches();

        // 过滤出适用于该游戏且未被禁用的补丁
        return installedPatches.stream()
                .filter(patch -> isPatchApplicableToGame(patch, gameId))
                .filter(patch -> config.isPatchEnabled(gameAsmPath, patch.manifest.id))
                .sorted(Comparator.comparingInt(patch -> -patch.manifest.priority))
                .collect(Collectors.toCollection(ArrayList::new));
    }

    /**
     * Returns all patches that are applicable to the specified game, regardless of enabled status.
     * Results are sorted by priority in descending order (higher priority first).
     *
     * @param gameId the game ID to get applicable patches for
     * @return list of all applicable patches, sorted by priority (highest first)
     */
    public ArrayList<Patch> getApplicablePatches(String gameId) {
        var installedPatches = getInstalledPatches();

        return installedPatches.stream()
                .filter(patch -> isPatchApplicableToGame(patch, gameId))
                .sorted(Comparator.comparingInt(patch -> -patch.manifest.priority))
                .collect(Collectors.toCollection(ArrayList::new));
    }

    /**
     * Check if a patch is applicable to a specific game.
     * Supports wildcard "*" in targetGames to match all games.
     *
     * @param patch  the patch to check
     * @param gameId the game ID to match against
     * @return true if the patch is applicable to the game
     */
    private boolean isPatchApplicableToGame(Patch patch, String gameId) {
        if (patch.manifest.targetGames == null || patch.manifest.targetGames.isEmpty()) {
            return false;
        }
        // Support wildcard "*" to match all games
        return patch.manifest.targetGames.contains("*") 
                || patch.manifest.targetGames.contains(gameId);
    }

    /**
     * Returns all patches that are enabled for the specified game assembly path.
     * Results are sorted by priority in descending order (higher priority first).
     * 
     * Note: By default, ALL installed patches are enabled. Only explicitly disabled
     * patches will be filtered out.
     *
     * @param gameAsmPath the game assembly path used to determine enabled patches
     * @return list of enabled patches, sorted by priority (highest first)
     */
    public ArrayList<Patch> getEnabledPatches(Path gameAsmPath) {
        var installedPatches = getInstalledPatches();
        
        // 过滤出未被禁用的补丁（默认全部启用）
        return installedPatches.stream()
                .filter(patch -> config.isPatchEnabled(gameAsmPath, patch.manifest.id))
                .sorted(Comparator.comparingInt(patch -> -patch.manifest.priority))
                .collect(Collectors.toCollection(ArrayList::new));
    }

    /**
     * Get IDs of all patches that are enabled for the specified game assembly path.
     *
     * @param gameAsmPath the game assembly path used to lookup enabled patch IDs
     * @return an ArrayList containing enabled patch IDs (may be empty)
     */
    public ArrayList<String> getEnabledPatchIds(Path gameAsmPath) {
        return config.getEnabledPatchIds(gameAsmPath);
    }

    /**
     * Scans the patch storage directory and returns all currently installed (valid) patches.
     *
     * @return list of all installed patches
     * @throws RuntimeException if an I/O error occurs while listing the directory (wrapped from IOException)
     */
    public ArrayList<Patch> getInstalledPatches() {
        ArrayList<Patch> installedPatches;
        try (var pathsStream = Files.list(patchStoragePath)) {
            installedPatches = pathsStream
                    .filter(Files::isDirectory)
                    .map(Patch::fromPatchPath)
                    .filter(Objects::nonNull)
                    .collect(Collectors.toCollection(ArrayList::new));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return installedPatches;
    }

    /**
     * Returns patches from the installed patches that match the provided IDs.
     * The returned list preserves the order of the supplied {@code patchIds} list.
     *
     * @param patchIds list of patch IDs to retrieve
     * @return list of patches matching the specified IDs, ordered to follow {@code patchIds}
     */
    public ArrayList<Patch> getPatchesByIds(List<String> patchIds) {
        var installedPatches = getInstalledPatches();
        return installedPatches.stream()
                .filter(patch -> patchIds.contains(patch.manifest.id))
                // sort by patchIds order
                .sorted(Comparator.comparingInt(patch -> patchIds.indexOf(patch.manifest.id)))
                .collect(Collectors.toCollection(ArrayList::new));
    }

    /**
     * Construct the environment variable string for startup hooks from a list of patches.
     * The absolute paths of the entry assemblies of the patches are joined with ':' as the separator.
     *
     * @param patches list of patches to construct the environment variable from
     * @return colon-separated environment variable string containing entry assembly absolute paths
     */
    public static String constructStartupHooksEnvVar(List<Patch> patches) {
        return patches
                .stream()
                .map(p -> p.getEntryAssemblyAbsolutePath().toString())
                .collect(Collectors.joining(":"));
    }

    //endregion

    //region Patch Installation

    /**
     * Install a patch archive (ZIP/7z). If a patch with the same ID already exists,
     * it will be deleted and the new archive will be extracted into the patch directory.
     * The method validates the patch manifest before installation.
     *
     * @param patchZipPath path to the patch archive file to install
     * @return true if installation succeeds, false otherwise
     */
    public boolean installPatch(Path patchZipPath) {
        if (!Files.exists(patchZipPath) || !Files.isRegularFile(patchZipPath)) {
            Log.w(TAG, "补丁安装失败: 补丁文件不存在或不是一个有效的文件, path: " + patchZipPath);
            return false;
        }

        var manifest = PatchManifest.fromZip(patchZipPath);
        if (manifest == null) {
            Log.w(TAG, "补丁安装失败: 无法读取补丁清单, path: " + patchZipPath);
            return false;
        }

        var patchPath = patchStoragePath.resolve(manifest.id);

        if (Files.exists(patchPath)) {
            Log.i(TAG, "补丁已存在, 将删除原补丁目录，重新安装, patch id: " + manifest.id);
            if (!FileUtils.deleteDirectoryRecursively(patchPath)) {
                Log.w(TAG, "删除原补丁目录时发生错误");
                return false;
            }
        } else {
            Log.i(TAG, "正在安装新补丁, patch id: " + manifest.id);
        }

        Log.i(TAG, "正在解压补丁文件到补丁目录...");
        new BasicSevenZipExtractor(patchZipPath, Paths.get(""), patchPath, new ExtractorCollection.ExtractionListener() {
            @Override
            public void onProgress(String message, float progress, HashMap<String, Object> state) {}

            @Override
            public void onComplete(String message, HashMap<String, Object> state) {}

            @Override
            public void onError(String message, Exception ex, HashMap<String, Object> state) {
                throw new RuntimeException(message, ex);
            }
        })
                .extract();

        return true;
    }

    //endregion

    //region Config Value Setting and Getting

    /**
     * Set whether a patch is enabled for a specific game.
     *
     * @param gameAsmPath the game assembly path
     * @param patchId     the patch ID
     * @param enabled     true to enable the patch, false to disable it
     */
    public void setPatchEnabled(Path gameAsmPath, String patchId, boolean enabled) {
        config.setPatchEnabled(gameAsmPath, patchId, enabled);
        saveConfig();
    }

    /**
     * Check if a patch is enabled for a specific game.
     *
     * @param gameAsmPath the game assembly path
     * @param patchId     the patch ID
     * @return true if the patch is enabled for the provided game assembly path, false otherwise
     */
    public boolean isPatchEnabled(Path gameAsmPath, String patchId) {
        return config.isPatchEnabled(gameAsmPath, patchId);
    }

    //endregion

    //region Configuration Management

    /**
     * Load configuration from file. If file doesn't exist or loading fails, creates a new config.
     */
    private void loadConfig() {
        config = PatchManagerConfig.fromJson(configFilePath);
        if (config == null) {
            Log.i(TAG, "配置文件不存在或加载失败，创建新配置");
            config = new PatchManagerConfig();
            saveConfig();
        } else {
            Log.i(TAG, "配置文件加载成功");
        }
    }

    /**
     * Save current configuration to file.
     */
    private void saveConfig() {
        if (!config.saveToJson(configFilePath)) {
            Log.w(TAG, "保存配置文件失败");
        }
    }

    //endregion

    //region Patch Storage Management

    /**
     * Get the default patch storage directory.
     * If {@code customStoragePath} is null, uses either the external files directory
     * or the internal files directory based on {@code IS_DEFAULT_PATCH_STORAGE_DIR_EXTERNAL}.
     *
     * @param customStoragePath custom storage path, or {@code null} to use the default location
     * @return path to the prepared patch storage directory
     * @throws IOException if directory creation fails
     */
    private static Path getDefaultPatchStorageDirectories(@Nullable String customStoragePath) throws IOException {
        // Implementation to prepare patch storage directories
        String baseDir;
        baseDir = Objects.requireNonNullElseGet(customStoragePath, () -> IS_DEFAULT_PATCH_STORAGE_DIR_EXTERNAL
                ? Objects.requireNonNull(Shared.getContext().getExternalFilesDir(null)).getAbsolutePath()
                : Shared.getContext().getFilesDir().getAbsolutePath());

        return Paths.get(baseDir, PATCH_STORAGE_DIR).normalize();
    }

    public static void installBuiltInPatches(PatchManager patchManager) {
        installBuiltInPatches(patchManager, false);
    }

    public static void installBuiltInPatches(PatchManager patchManager, boolean forceReinstall) {
        Path apkPath = Paths.get(Shared.getContext().getApplicationInfo().sourceDir);
        try (var tfa = new TemporaryFileAcquirer()) {
            Path extractedPatches = tfa.acquireTempFilePath("extracted_patches");

            new BasicSevenZipExtractor(
                    apkPath,
                    Paths.get("assets/patches"),
                    extractedPatches,
                    new ExtractorCollection.ExtractionListener() {
                        @Override
                        public void onProgress(String message, float progress, HashMap<String, Object> state) {}

                        @Override
                        public void onComplete(String message, HashMap<String, Object> state) {}

                        @Override
                        public void onError(String message, Exception ex, HashMap<String, Object> state) {
                            throw new RuntimeException(message, ex);
                        }
                    }
            )
                    .extract();

            // 安装共享依赖库（如 0Harmony.dll）
            try (var pathsStream = Files.list(extractedPatches)) {
                var dlls = pathsStream
                        .filter(path -> Files.isRegularFile(path) && path.toString().endsWith(".dll"))
                        .collect(Collectors.toList());
                for (var dll : dlls) {
                    Path targetPath = patchManager.patchStoragePath.resolve(dll.getFileName());
                    if (!Files.exists(targetPath)) {
                        Log.i(TAG, "正在安装补丁依赖库: " + dll.getFileName());
                        Files.copy(dll, targetPath);
                    }
                }
            } catch (IOException e) {
                throw new RuntimeException(e);
            }

            // 获取已安装补丁的 ID 列表
            var installedPatchIds = patchManager.getInstalledPatches().stream()
                    .map(p -> p.manifest.id)
                    .collect(Collectors.toSet());

            // 安装缺失的内置补丁
            try (var pathsStream = Files.list(extractedPatches)) {
                var patchZips = pathsStream
                        .filter(path -> Files.isRegularFile(path) && path.toString().endsWith(".zip"))
                        .collect(Collectors.toList());
                for (var patchZip : patchZips) {
                    // 读取补丁 manifest 检查是否已安装
                    PatchManifest manifest = PatchManifest.fromZip(patchZip);
                    if (manifest != null && forceReinstall) {
                        Log.i(TAG, "正在强制重新安装内置补丁: " + patchZip.getFileName() + " (id: " + manifest.id + ")");
                        patchManager.installPatch(patchZip);
                    } else if (manifest != null && !installedPatchIds.contains(manifest.id)) {
                        Log.i(TAG, "正在安装内置补丁: " + patchZip.getFileName() + " (id: " + manifest.id + ")");
                        patchManager.installPatch(patchZip);
                    } else if (manifest != null) {
                        Log.d(TAG, "补丁已安装，跳过: " + manifest.id);
                    }
                }
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
    }

    //endregion
}

File: app/src/main/java/com/app/ralaunch/feature/patch/data/PatchManager.kt
package com.app.ralaunch.feature.patch.data

import android.content.Context
import android.util.Log
import com.app.ralaunch.core.platform.install.extractors.BasicSevenZipExtractor
import com.app.ralaunch.core.platform.install.extractors.ExtractorCollection
import com.app.ralaunch.core.common.util.FileUtils
import com.app.ralaunch.core.common.util.TemporaryFileAcquirer
import org.koin.java.KoinJavaComponent
import java.io.File
import java.io.IOException

class PatchManager @JvmOverloads constructor(
    customStoragePath: String? = null,
    installPatchesImmediately: Boolean = false
) {
    private var patchStorageDir: File? = null
    private var configFile: File? = null
    private var config: PatchManagerConfig = PatchManagerConfig()

    init {
        try {
            val dir = getDefaultPatchStorageDirectories(customStoragePath)
            patchStorageDir = dir
            
            if (!dir.exists() || !dir.isDirectory) {
                dir.deleteRecursively()
                dir.mkdirs()
            }
            
            val cfgFile = File(dir, PatchManagerConfig.CONFIG_FILE_NAME)
            configFile = cfgFile
            config = loadConfig(cfgFile)

            cleanLegacySharedDlls(dir)

            if (installPatchesImmediately) {
                installBuiltInPatches(this)
            }
        } catch (e: Exception) {
            Log.e(TAG, "FATAL ERROR during PatchManager initialization: ${e.message}", e)
            // We swallow the error so the app doesn't crash.
            // Patch features will just be disabled.
        }
    }

    fun getApplicableAndEnabledPatches(gameId: String, gameAsmPath: File): ArrayList<Patch> {
        return installedPatches
            .filter { isPatchApplicableToGame(it, gameId) }
            .filter { config.isPatchEnabled(gameAsmPath, it.manifest.id) }
            .sortedByDescending { it.manifest.priority }
            .toCollection(ArrayList())
    }

    fun getApplicablePatches(gameId: String): ArrayList<Patch> {
        return installedPatches
            .filter { isPatchApplicableToGame(it, gameId) }
            .sortedByDescending { it.manifest.priority }
            .toCollection(ArrayList())
    }

    private fun isPatchApplicableToGame(patch: Patch, gameId: String): Boolean {
        val targetGames = patch.manifest.targetGames
        if (targetGames.isNullOrEmpty()) return false
        return targetGames.contains("*") || targetGames.contains(gameId)
    }

    fun getEnabledPatches(gameAsmPath: File): ArrayList<Patch> {
        return installedPatches
            .filter { config.isPatchEnabled(gameAsmPath, it.manifest.id) }
            .sortedByDescending { it.manifest.priority }
            .toCollection(ArrayList())
    }

    fun getEnabledPatchIds(gameAsmPath: File): ArrayList<String> {
        return config.getEnabledPatchIds(gameAsmPath)
    }

    val installedPatches: ArrayList<Patch>
        get() {
            val dir = patchStorageDir ?: return ArrayList()
            return try {
                dir.listFiles()
                    ?.filter { it.isDirectory }
                    ?.mapNotNull { Patch.fromPatchPath(it) }
                    ?.toCollection(ArrayList())
                    ?: ArrayList()
            } catch (e: Exception) {
                Log.e(TAG, "Error reading installed patches: ${e.message}")
                ArrayList()
            }
        }

    fun getPatchesByIds(patchIds: List<String>): ArrayList<Patch> {
        return installedPatches
            .filter { patchIds.contains(it.manifest.id) }
            .sortedBy { patchIds.indexOf(it.manifest.id) }
            .toCollection(ArrayList())
    }

    fun installPatch(patchZipFile: File): Boolean {
        val storageDir = patchStorageDir ?: return false
        
        if (!patchZipFile.exists() || !patchZipFile.isFile) {
            Log.w(TAG, "Patch install failed: file not found: $patchZipFile")
            return false
        }

        val manifest = PatchManifest.fromZip(patchZipFile)
        if (manifest == null) {
            Log.w(TAG, "Patch install failed: cannot read manifest: $patchZipFile")
            return false
        }

        val patchDir = File(storageDir, manifest.id)

        if (patchDir.exists()) {
            Log.i(TAG, "Patch exists, reinstalling: ${manifest.id}")
            if (!FileUtils.deleteDirectoryRecursively(patchDir)) {
                Log.w(TAG, "Failed to delete old patch dir")
                return false
            }
        } else {
            Log.i(TAG, "Installing new patch: ${manifest.id}")
        }

        Log.i(TAG, "Extracting patch...")
        return try {
            BasicSevenZipExtractor(
                patchZipFile,
                File(""),
                patchDir,
                object : ExtractorCollection.ExtractionListener {
                    override fun onProgress(message: String, progress: Float, state: HashMap<String, Any?>?) {}
                    override fun onComplete(message: String, state: HashMap<String, Any?>?) {}
                    override fun onError(message: String, ex: Exception?, state: HashMap<String, Any?>?) {
                        Log.e(TAG, "Extraction error: $message", ex)
                    }
                }
            ).extract()
        } catch (e: Exception) {
            Log.e(TAG, "Failed to extract patch", e)
            false
        }
    }

    fun setPatchEnabled(gameAsmPath: File, patchId: String, enabled: Boolean) {
        config.setPatchEnabled(gameAsmPath, patchId, enabled)
        saveConfig()
    }

    fun isPatchEnabled(gameAsmPath: File, patchId: String): Boolean {
        return config.isPatchEnabled(gameAsmPath, patchId)
    }

    private fun loadConfig(cfgFile: File): PatchManagerConfig {
        return try {
            val loadedConfig = PatchManagerConfig.fromJson(cfgFile)
            if (loadedConfig == null) {
                Log.i(TAG, "Config not found, creating new")
                val newCfg = PatchManagerConfig()
                newCfg.saveToJson(cfgFile)
                newCfg
            } else {
                Log.i(TAG, "Config loaded")
                loadedConfig
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error loading config: ${e.message}")
            PatchManagerConfig() // Return default config on crash
        }
    }

    private fun saveConfig() {
        val cfgFile = configFile ?: return
        if (!config.saveToJson(cfgFile)) {
            Log.w(TAG, "Failed to save config")
        }
    }

    private fun cleanLegacySharedDlls(dir: File) {
        for (dllName in LEGACY_SHARED_DLLS) {
            val dllFile = File(dir, dllName)
            try {
                if (dllFile.exists()) {
                    dllFile.delete()
                    Log.i(TAG, "Cleaned legacy DLL: $dllName")
                }
            } catch (e: Exception) {
                Log.w(TAG, "Failed to clean $dllName: ${e.message}")
            }
        }
    }

    companion object {
        private const val TAG = "PatchManager"
        private const val IS_DEFAULT_PATCH_STORAGE_DIR_EXTERNAL = true
        private const val PATCH_STORAGE_DIR = "patches"

        private val LEGACY_SHARED_DLLS = arrayOf(
            "0Harmony.dll",
            "MonoMod.Common.dll",
            "Mono.Cecil.dll"
        )

        private fun getDefaultPatchStorageDirectories(customStoragePath: String?): File {
            val context: Context = KoinJavaComponent.get(Context::class.java)
            val baseDir = customStoragePath ?: if (IS_DEFAULT_PATCH_STORAGE_DIR_EXTERNAL) {
                // FIXED: Safely handle null externalFilesDir
                val extDir = context.getExternalFilesDir(null)
                extDir?.absolutePath ?: context.filesDir.absolutePath
            } else {
                context.filesDir.absolutePath
            }
            
            // FIXED: Removed canonicalFile. If path has symlinks, canonicalFile crashes on some Android 7 devices.
            return File(baseDir, PATCH_STORAGE_DIR).absoluteFile
        }

        @JvmStatic
        fun installBuiltInPatches(patchManager: PatchManager) {
            installBuiltInPatches(patchManager, false)
        }

        @JvmStatic
        fun installBuiltInPatches(patchManager: PatchManager, forceReinstall: Boolean) {
            try {
                val context: Context = KoinJavaComponent.get(Context::class.java)
                val apkFile = File(context.applicationInfo.sourceDir)

                TemporaryFileAcquirer().use { tfa ->
                    val extractedPatches = tfa.acquireTempFilePath("extracted_patches")

                    BasicSevenZipExtractor(
                        apkFile,
                        File("assets/patches"),
                        extractedPatches,
                        object : ExtractorCollection.ExtractionListener {
                            override fun onProgress(message: String, progress: Float, state: HashMap<String, Any?>?) {}
                            override fun onComplete(message: String, state: HashMap<String, Any?>?) {}
                            override fun onError(message: String, ex: Exception?, state: HashMap<String, Any?>?) {
                                Log.e(TAG, "Built-in extraction error: $message", ex)
                            }
                        }
                    ).extract()

                    val installedPatchMap = patchManager.installedPatches
                        .associateBy { it.manifest.id }

                    extractedPatches.listFiles()
                        ?.filter { it.isFile && it.name.endsWith(".zip") }
                        ?.forEach { patchZip ->
                            try {
                                val manifest = PatchManifest.fromZip(patchZip)
                                    ?: return@forEach

                                val installedPatch = installedPatchMap[manifest.id]

                                when {
                                    forceReinstall -> {
                                        Log.i(TAG, "Force reinstalling: ${patchZip.name}")
                                        patchManager.installPatch(patchZip)
                                    }
                                    installedPatch == null -> {
                                        Log.i(TAG, "Installing: ${patchZip.name}")
                                        patchManager.installPatch(patchZip)
                                    }
                                    else -> {
                                        val installedVersion = installedPatch.manifest.version
                                        val bundledVersion = manifest.version
                                        val cmp = PatchManifest.compareVersions(bundledVersion, installedVersion)
                                        if (cmp > 0) {
                                            Log.i(TAG, "Updating patch: ${manifest.id} ($installedVersion -> $bundledVersion)")
                                            patchManager.installPatch(patchZip)
                                        } else {
                                            Log.d(TAG, "Patch up to date: ${manifest.id}")
                                        }
                                    }
                                }
                            } catch (e: Exception) {
                                Log.e(TAG, "Error installing individual patch: ${patchZip.name}", e)
                            }
                        }
                }
            } catch (e: Exception) {
                Log.e(TAG, "FATAL ERROR during built-in patch installation: ${e.message}", e)
            }
        }

        @JvmStatic
        fun constructStartupHooksEnvVar(patches: List<Patch>): String {
            return try {
                val seenPatchIds = linkedSetOf<String>()
                patches
                    .filter { seenPatchIds.add(it.manifest.id) }
                    .map { it.getEntryAssemblyAbsolutePath().absolutePath } // FIXED: toString -> absolutePath
                    .distinct()
                    .joinToString(":")
            } catch (e: Exception) {
                Log.e(TAG, "Error constructing startup hooks", e)
                ""
            }
        }
    }
}

